lib/Mojo/WebSocketProxy/Dispatcher.pm

```diff
@@ -66,40 +72,68 @@ sub on_message {
 
     my $config = $c->wsp_config->{config};
 
-    my $result;
     my $req_storage = {};
     $req_storage->{args} = $args;
-    timeout 15 => sub {
-        $result = Mojo::WebSocketProxy::Parser::parse_req($c, $req_storage);
-        if (!$result
-            && (my $action = $c->dispatch($args)))
-        {
-            %$req_storage = (%$req_storage, %$action);
-            $req_storage->{method} = $req_storage->{name};
-            $result = $c->before_forward($req_storage);
-
-            # Don't forward call to RPC if any before_forward hook returns response
-            # Or if there is instead_of_forward action
-            unless ($result) {
-                $result =
-                      $req_storage->{instead_of_forward}
-                    ? $req_storage->{instead_of_forward}->($c, $req_storage)
-                    : $c->forward($req_storage);
-            }
-
-            $c->after_forward($result, $req_storage);
-        } elsif (!$result) {
-            $c->app->log->debug("unrecognised request: " . $c->dumper($args));
-            $result = $c->wsp_error('error', 'UnrecognisedRequest', 'Unrecognised request.');
+
+    my $result = Mojo::WebSocketProxy::Parser::parse_req($c, $req_storage);
```

Comment:
It is not assumed, that the `parse_req` should be **asynchronous**. That's why `parse_req` should not return the Future. The `$result` is wrapped into future here, a few lines below.